### DISCUSSION TITLE
How to stitch images containing white space

### DISCUSSION THREAD

**Discussion by Greatuso** [OpenStitching/stitching#172] [Q&A] [answered]
Hi,
Thanks for amazing your work!But now I encountered a problem about how to stitch blank areas.
<img width="356" alt="awhite" src="https://github.com/OpenStitching/stitching/assets/47665698/c0441ac3-1b21-47ef-ba9b-39f535564db3">
As shown in the above figure，2-6.jpg,3-6.jpg,4-6.jpg，they don't have enough features to match.How should I do  to keep the output complete？I would greatly appreciate it if you could answer.

**Comment by lukasalexanderweber** [comment]
What happens if you set crop to false? The images should then be ignored and you should get a black area where they are, which you could handle in some preprocessing

**Reply by lukasalexanderweber** [reply]
*postprocessing

**Reply by Greatuso** [reply]
Thanks for your answer.I  took your suggestion and made an attempt.

Although I set crop to false,it won't stitch all images.For example, it stitched some pictures on the left side,but about the blank areas it would prompt 'Not all images are included in the final panorama' and 'You might want to lower the confidence_threshold or try another detector to include all your images'.

I want to stitch all images in a panorama,I mean when there are enough features I blend the images, if there are not enough features, I stitch them according to their positions.Its working mode looks like a pathological scanner,I've got the stitching order and the coverage area between images.

Thank you again for your guidance

**Reply by lukasalexanderweber** [reply]
Can you share all Images?

**Reply by Greatuso** [reply]
Yeah, I will collect the data in a more standardized manner and send the package to you later.

**Comment by Greatuso** [comment]
The File size is a little big.I need to divide them into three parts and upload them.The first package is data-1.zip,the second and the third are the data-2.zip
[Uploading data-1.zip…]()

**Reply by Greatuso** [reply]
The data-1.zip is still uploading,if it failed,I'll upload it later

**Comment by Greatuso** [comment]
[data-2(1).zip](https://github.com/OpenStitching/stitching/files/13374471/data-2.1.zip)
[data-2(2).zip](https://github.com/OpenStitching/stitching/files/13374472/data-2.2.zip)
[data-2(3).zip](https://github.com/OpenStitching/stitching/files/13374476/data-2.3.zip)
 This is  data-2.zip.
when I get the panorama,I want to view the complete organization,including its spatial location to distinguish which one it is.So I need to stitch the blank area. The data-2.zip contains a whole slide.
<img width="1218" alt="b801674f808aadaf15bbe029c4e432b" src="https://github.com/OpenStitching/stitching/assets/47665698/b73291a9-2b41-47c1-be9d-0e02de1a370b">
The above image  was stitch using SIFT before, and I forcefully overlaid it when there were not enough features.It seems to have many drawbacks.
When I find your repo,it really suprises me.It not only solved the problem of exposure compensation, but also integrated it more smoothly.So I want to apply it to my project.
Thanks again!

**Comment by Greatuso** [comment]
[data-1.zip](https://github.com/OpenStitching/stitching/files/13385456/data-1.zip)

**Reply by lukasalexanderweber** [reply, marked as answer]
I get this result with `stitch * -v --affine --no-crop`.

I don't know how the images are taken, but I expected `affine` relationship.

I guess this is the best you can get. I would suggest to replace black color pixels with the cells background color.

![09_result](https://github.com/OpenStitching/stitching/assets/32765578/eddde53b-19d6-471b-bc5a-2cc2df82165f)

**Reply by Greatuso** [reply]
Oh thanks，I'll try it later，by the way,did you stitch all the images together? Or did you stitch them together row by row first and then join all the rows together?
Another question I have is, was cylindrical projection used in the final image?

**Reply by lukasalexanderweber** [reply]
All at once and with affine projection

**Reply by Greatuso** [reply]
I put all images and I meet  "stitching.stitching_error.StitchingError: Camera parameters adjusting failed."
Did you only use the 'stitch * -v --affine --no-crop' command?
Thanks again!

**Reply by Greatuso** [reply]
![panorama0](https://github.com/OpenStitching/stitching/assets/47665698/eab0debf-ccdb-42fa-a43d-3b21032c183d)
![panorama111](https://github.com/OpenStitching/stitching/assets/47665698/11da7514-efc0-4592-9511-d946d981b11d)
![panorama222](https://github.com/OpenStitching/stitching/assets/47665698/e561280a-5038-4dbd-a10e-8cd64303a846)
When I stitch panorama111&panorama222，the panorama0 is the result.The result has undergone deformation. Why is this and how can I avoid it
panowz.jpg is the result used '--affine'
![panowz](https://github.com/OpenStitching/stitching/assets/47665698/5292681d-c8c2-452d-940f-faafd4ef499a)

**Reply by Greatuso** [reply]
--wave_correct_kind no. I find it from issue.And I put it and solved the question. Thanks

**Reply by lukasalexanderweber** [reply]
thats strange: `--wave_correct_kind no` should automatically be set if `--affine` is used, see [AffineStitcher](https://github.com/OpenStitching/stitching/blob/main/stitching/stitcher.py#L265)

**Comment by Greatuso** [comment]
Hi，I have met another problem。
<img width="713" alt="img1" src="https://github.com/OpenStitching/stitching/assets/47665698/6edbffcb-a956-4968-9b95-cce1a5542952">
<img width="716" alt="img2" src="https://github.com/OpenStitching/stitching/assets/47665698/3b7b094b-9a5a-469b-b008-164542c679b7">
I failed to stitch them together. The red box is their overlapping area, and it is obvious that the feature points outside the red box may have interfered with the stitching. What should I do to make feature detection only detect red boxes? Or perhaps I have read '#55' but what still confuses me is how to input the position/coordinates of the mask.
Thanks！

**Comment by lukasalexanderweber** [comment]
See 

https://github.com/OpenStitching/stitching_tutorial/blob/master/docs/Stitching%20Tutorial.md#feature-masking 

and

https://github.com/OpenStitching/stitching_tutorial/tree/master/imgs (Barcode, Mask)

**Reply by Greatuso** [reply]
Thanks!

**Comment by Greatuso** [comment]
![image](https://github.com/OpenStitching/stitching/assets/47665698/c8a70b57-5381-4b3d-b9f7-7cd110b49174)
![image](https://github.com/OpenStitching/stitching/assets/47665698/427e30f0-eb65-486d-bfae-3db7b8d15742)

I found that my image is a bit blurry, and some details will be lost after stitiching.But your image always retains details.
As shown in the above picture, the cells in my picture have become blurred, while the cells in your picture are still clear. I want to know why this is happening?Thank you!

**Comment by lukasalexanderweber** [comment]
Are you stitching them one after another as https://github.com/OpenStitching/stitching/discussions/84

**Reply by Greatuso** [reply]
As you said,I stitched them one after another one.
I'm afraid that I  could not  collect all Images in a List and stitch them together .The working mode of the camera is like a pathological scanner,so I have to stitch them one by one.
![image](https://github.com/OpenStitching/stitching/assets/47665698/5195b01a-6a3f-45a0-88b2-d15f47122f20)
The panorama is like this. The data I previously shared is a part of the panorama .It may contain thousands of images .

**Comment by Greatuso** [comment]
Hi，when I put all the images into Stitcher at once, this error occurred:
"cv2.error: OpenCV(4.8.1) D:\a\opencv-python\opencv-python\opencv\modules\flann\src\miniflann.cpp:521: error: (-215:Assertion failed) (size_t)knn <= index_->size() in function 'cv::flann::runKnnSearch_'
"
The images contain cells and blank areas.It is easy to occur when there are many blank areas
![image](https://github.com/OpenStitching/stitching/assets/47665698/ca12dd1a-dbf5-4b2b-8d61-4be065daa779)
As shown in the figure, the problem does not occur when I select the red area, but it does occur when I select the blue area.
Have you ever encountered this problem before? Could you please guide me on how to correct it?
